clang/include/clang/Basic/BuiltinsNVPTX.def

```diff
@@ -650,6 +663,1378 @@ BUILTIN(__nvvm_atom_cas_gen_ll, "LLiLLiD*LLiLLi", "n")
+TARGET_BUILTIN(__nvvm_atom_acq_rel_dec_gen_ull, "ULLiULLiD*ULLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_dec_gen_ull, "ULLiULLiD*ULLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_dec_gen_ull, "ULLiULLiD*ULLi", "n", SM_70)
+
+TARGET_BUILTIN(__nvvm_atom_acq_rel_and_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_and_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_and_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_and_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_and_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_and_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_and_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_and_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_and_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+
+TARGET_BUILTIN(__nvvm_atom_acq_rel_or_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_or_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_or_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_or_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_or_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_or_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_or_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_or_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_or_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+
+TARGET_BUILTIN(__nvvm_atom_acq_rel_xor_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_xor_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_xor_gen_i, "iiD*i", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_xor_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_xor_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_xor_gen_l, "LiLiD*Li", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_xor_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_xor_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_xor_gen_ll, "LLiLLiD*LLi", "n", SM_70)
+
+TARGET_BUILTIN(__nvvm_atom_acq_rel_cas_gen_i, "iiD*ii", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_cas_gen_i, "iiD*ii", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_cas_gen_i, "iiD*ii", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_cas_gen_l, "LiLiD*LiLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_cas_gen_l, "LiLiD*LiLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_cas_gen_l, "LiLiD*LiLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_acq_rel_cas_gen_ll, "LLiLLiD*LLiLLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_cta_acq_rel_cas_gen_ll, "LLiLLiD*LLiLLi", "n", SM_70)
+TARGET_BUILTIN(__nvvm_atom_sys_acq_rel_cas_gen_ll, "LLiLLiD*LLiLLi", "n", SM_70)
+
+BUILTIN(__nvvm_atom_add_global_i, "iiD*i", "n")
```

Comment:
I cannot help but wonder if the builtins could select the right intrinsics without the space explicitly stated in the builtin's name, i.e. inferring it from the pointer's address space. Or maybe even as late as instruction selection?
@Naghasan - Do you have insight into whether or not that is possible?